### ISSUE TITLE
Preprocessor support

### ISSUE THREAD

**Issue by vamshikrishna83** [opentofu/opentofu#2130]
### OpenTofu Version

```shell
OpenTofu v1.8.4
on linux_amd64
```

### The problem in your OpenTofu project

We use same resource configuration to be deployed across multiple environments like development, staging and production.
Obviously, there would be some minor enhancements that only need to be deployed to development / staging but not to production environments. With the current feature set, it is not possible to acheive this behaviour.

### Attempted Solutions

We tried to add a child module which includes the specific configuration to be deployed only to a sepcific environment, and conditionally include this child module to be processed only for development / staging environments. But, this approach doesn't work well - if we need to make updates to existing resources being configured as part of this run. Like add an new route in existing APIGW configuration, which can't be done easily.

### Proposal

Enable c style preprocessor or even better to add ability to conditionally execute a resource. Similar functionality has been added as part of [terracotta](https://github.com/toddlucas/terracotta) utility for terraform, but would be nice to include the same and extend it in opentofu.

```hcl
!if ENVIRONMENT == "development"
resource "null_resource" "null" {
}
!endif
``` 

or 

```hcl
!if CONDITION
resource "null_resource" "null" {
}
!endif
```

**Comment by cam72cam** [comment]
Closing as duplicate of #2129

**Comment by vamshikrishna83** [comment]
May be i chose a simple example, that doesn't explain our case properly. Let me try again with an actual use-case we have today.

We try to configure routes on OCI Load Balancer and APIGW back to our backends, and as part of it - we need to add new routes / update existing routes as and when we make new changes. These new changes will initially be deployed only to development environments, and validated - after testing and stabilization, these changes will be upgraded / marked to be deployed to production.

For example, in the below example `/existing` is an existing route, and `/new` is a new route being added to the APIGW deployment, which needs to be deployed only in development environments until testing is completed.

```hcl
resource "oci_apigateway_deployment" "existing_deployment" {
  compartment_id = var.COMPARTMENT
  gateway_id     = var.API_GATEWAY_ID
  path_prefix    = "/v2"
  display_name   = "deployment"

  specification {
   # Existing route in APIGW deployment
    routes {
      path = "/existing"
      backend {
        type   = "STOCK_RESPONSE_BACKEND"
        body   = "Stock Response Succeeded."
      }
      methods = ["ANY"]
    }
   !if ENVIRONMENT == "development"
    routes {
      path = "/new"
      backend {
        type   = "STOCK_RESPONSE_BACKEND"
        status = 200
        body   = "Stock Response Succeeded."
      }
      methods = ["ANY"]
    }
   !endif
  }
}
``` 

Unfortunately, it is not possible to implement this using `count` option. `count` option is supported at a resource level, where as the current situation, we need this sort of control at a sub-resource level which is not supported.

Once the verification of the new route is done, then all we need to do is to get rid of the preprocessor `!if` conditional, which makes these changes to be deployed for all environments. Hope this example, makes it clear on the need for such a preprocessor construct.

**Comment by vamshikrishna83** [comment]
Yes, we have considered and using this approach for some time now. There are some complexities in making this dynamic block work in general for all possible route combinations.

There are more than 50+ properties available as part of route block, and to configure each value generically for all possible usable fields and to leave in valid default values for other un-used fields will make it really complex and not easy to use.

Last but not least, once we validate the new routes in development environment - switching from dynamic block back to route block is not trivial enough - and would require some more testing to validate no regressions are introduced as part of this conversion.

**Comment by vamshikrishna83** [comment]
Thanks a lot for your detailed explaination.

As of now, we are using dynamic blocks to make it work. We are open to try out other options that might make our lives easier in this regard.  With our basic understanding, we felt pre-processor / override files are two other ways we could explore - hence this issue. If the `count` conditional support is available at a sub-resource level, then we are willing to consider that as well.

Thanks.
